module/zfs/vdev.c

```diff
@@ -3670,6 +3671,116 @@ vdev_deadman(vdev_t *vd)
 	}
 }
 
+/*
+ * Implements the per-vdev portion of manual TRIM. The function passes over
+ * all metaslabs on this vdev and performs a metaslab_trim_all on them. It's
+ * also responsible for rate-control if spa_man_trim_rate is non-zero.
+ *
+ * If fulltrim is set, metaslabs without spacemaps are also trimmed.
```

Comment:
This was done in order to be more friendly to certain storage backends that may be very sparsely allocated and have a very high cost for trim operations as suggested by @gaurkuma.  It seemed easy enough and pretty harmless to implement.